sematic/resolvers/log_streamer.py

```diff
@@ -12,36 +12,106 @@
+    is the one capturing the stdout of its child. We want the child to capture the
+    stdout of its parent. Why? Because we want the parent process to represent the "real"
+    work. There are a few reasons for this. One is so that the exit code for the parent
+    process represents the exit code for the real work. Another is so that if other
+    infra (ex: k8s) sends signals to the parent process, they reach the process doing
+    the "real work" first, which means there is a lower chance of weird issues with
+    forwarding that signal to children and cleaning up properly afterwards. It also
+    just makes more sense from a mental-model standpoint to have the daemon-like
+    process be the child while the "parent" process and the "main" process are the
+    same.
 """
 
 
 DEFAULT_LOG_UPLOAD_INTERVAL_SECONDS = 10
+_LAST_NON_EMPTY_DELTA_TEMPLATE = "{}.previous"
+
 
+def _flush_to_file(file_path, read_handle, uploader, remote_prefix, timeout_seconds=None):
+    """Read from the read_handle dump to file_path and then remote.
 
-def _stream_logs_to_remote_from_file(
+    The read_handle is continuously streamed from onto disk. The remote upload will
+    only happen once (a) the timeout is reached OR (b) there are no more contents
+    from the read handle. Once a remote upload has been performed, this exits.
+
+    Parameters
+    ----------
+    file_path:
+        Path of the file to stream the read_handle contents to. This will contain the
+        log file "delta"
+    read_handle:
+        A readable object that can be streamed from. It should be configured such that
+        reads are non-blocking.
+    uploader:
+        The function to call for performing the remote upload
+    remote_prefix:
+        The prefix for the remote storage location where the log files will be kept.
+        The actual file name will be unique for each upload, increasing monoatonically
+        with time
+    timeout_seconds:
+        The max number of seconds between when streaming from the read handle starts and
+        when the upload occurs.
+    """
+    if os.path.exists(file_path):
+        if os.stat(file_path)[stat.ST_SIZE] > 0:
+            # save the last non-empty delta file somewhere for tailing
+            os.rename(file_path, _LAST_NON_EMPTY_DELTA_TEMPLATE.format(file_path))
+        else:
+            os.remove(file_path)
+    started_reading = time.time()
+    # Use w+ mode; should overwrite whatever was in the prior delta file
+    with open(file_path, "w+") as fp:
+        while timeout_seconds is None or time.time() - started_reading < timeout_seconds:
+            line = read_handle.readline()
+            if len(line) > 0:
+                # The line would at least have the newline char if it was a blank.
+                # hence the "if"
+                fp.write(line)
+            else:
+                break  # no more to read right now; go ahead and flush
+            fp.flush()
+    uploader(file_path, remote_prefix)
```

Comment:
General nit, as you know I like airy code 😬 especially when it's hairy 🍝 

Can you add a blank line before `started_reading`, `# Use w+ mode`, and `uploader`?

Reply:
yup!